utility/session.go

```diff
@@ -0,0 +1,138 @@
+)
+
+type Session interface {
+	NewSession(sessionHeight int64, blockHash string, geoZone GeoZone, relayChain RelayChain, application modules.Actor) (Session, types.Error)
+	GetServiceNodes() []modules.Actor // the ServiceNodes providing Web3 to the application
+	GetFishermen() []modules.Actor    // the Fishermen monitoring the serviceNodes
+	GetApplication() modules.Actor    // the Application consuming the web3 access
+	GetRelayChain() RelayChain        // the chain identifier of the web3
+	GetGeoZone() GeoZone              // the geolocation zone where all are registered
+	GetSessionHeight() int64          // the block height when the session started
+}
+
+type RelayChain Identifier
+type GeoZone Identifier
+
+type Identifier interface {
+	Name() string
+	ID() string
+	Bytes() []byte
+}
+
+var _ Session = &session{}
+
+type session struct {
+	serviceNodes  []modules.Actor
+	fishermen     []modules.Actor
+	application   modules.Actor
+	relayChain    RelayChain
+	geoZone       GeoZone
+	blockHash     string
+	key           []byte
+	sessionHeight int64
+}
+
+func (s *session) NewSession(sessionHeight int64, blockHash string, geoZone GeoZone, relayChain RelayChain, application modules.Actor) (session Session, err types.Error) {
+	s.sessionHeight = sessionHeight
+	s.blockHash = blockHash
+	s.geoZone = geoZone
+	s.relayChain = relayChain
+	s.application = application
+	s.key, err = s.sessionKey()
+	if err != nil {
+		return
+	}
+	s.serviceNodes = s.findClosestXServiceNodes()
+	s.fishermen = s.findClosestYFishermen()
+	return s, nil
+}
+
+// use the seed information to determine a SHA3Hash that is used to find the closest N actors based
+// by comparing the sessionKey with the actors' public key
+func (s *session) sessionKey() ([]byte, types.Error) {
+	sessionHeightBytes := make([]byte, 8)
+	binary.LittleEndian.PutUint64(sessionHeightBytes, uint64(s.sessionHeight))
+	blockHashBz, err := hex.DecodeString(s.blockHash)
+	if err != nil {
+		return nil, types.ErrHexDecodeFromString(err)
+	}
+	appPubKey, err := crypto.NewPublicKey(s.application.GetPublicKey())
+	if err != nil {
+		return nil, types.ErrNewPublicKeyFromBytes(err)
+	}
+	return s.concat(sessionHeightBytes, blockHashBz, s.geoZone.Bytes(), s.relayChain.Bytes(), appPubKey.Bytes()), nil
+}
+
+// uses the current 'world state' to determine the service nodes in the session
+// 1) get an ordered list of the public keys of service nodes who are:
+//    - actively staked
+//    - staked within geo-zone
+//    - staked for relay-chain
+// 2) calls `pseudoRandomSelection(serviceNodes, numberOfNodesPerSession)`
+func (s *session) findClosestXServiceNodes() []modules.Actor {
+	// TODO (@andrewnguyen22) implement me
+	return nil
+}
+
+// uses the current 'world state' to determine the fishermen in the session
+// 1) get an ordered list of the public keys of fishermen who are:
+//    - actively staked
+//    - staked within geo-zone
+//    - staked for relay-chain
```

Comment:
Per the fisherman protocol [1], I didn't think that fish stake for a specific relayChain. Should they be agnostic?

[1] https://github.com/pokt-network/pocket-network-protocol/tree/main/utility#33-fisherman-protocol

Reply:
From the source you linked:

> Upon registration, a Fishermen must provide the network necessary information to interact with the protocol including StakeAmount, GeoZone, optional OperatorPubKey, and the Pocket API endpoint where the Fishermen conduct public services

Reply:
This is an interesting question and is blocked by [`V1 Scope Spike`  ](https://github.com/pokt-network/pocket-network-protocol/issues/18)

Fishermen are only going to be applicable to Pocket Supported Relay Chains (where the protocol pays out for the relay chain) 

It is unclear at this time what the limitations and scoping will be for Fishermen RelayChain support.

Reply:
Let's just add it to the Q&A for now and say it may change.